classes/invitation/models/Invitation.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @file classes/invitation/models/Invitation.php
+ *
+ * Copyright (c) 2014-2023 Simon Fraser University
+ * Copyright (c) 2000-2023 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class Invitation
+ *
+ * @brief Laravel Eloquent model for Invitation (access_keys table)
+ */
+
+namespace PKP\invitation\models;
+
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Support\InteractsWithTime;
+use PKP\invitation\invitations\enums\InvitationStatus;
+use PKP\invitation\traits\Attributes;
+use PKP\job\casts\DatetimeToInt;
+
+class Invitation extends Model
+{
+    use Attributes;
+    use InteractsWithTime;
+
+    /**
+     * Model's database table
+     *
+     * @var string
+     */
+    protected $table = 'access_keys';
+
+    /**
+     * Model's primary key
+     *
+     * @var string
+     */
+    protected $primaryKey = 'access_key_id';
+
+    /**
+     * Model's timestamp fields
+     *
+     * @var bool
+     */
+    public $timestamps = true;
+
+    /**
+     * The attributes that are not mass assignable.
+     *
+     * @var string[]|bool
+     */
+    protected $guarded = [];
+
+    /**
+     * Casting attributes to their native types
+     *
+     * @var string[]
+     */
+    protected $casts = [
+        'key_hash' => 'string',
+        'payload' => 'array',
+        'user_id' => 'int',
+        'assoc_id' => 'int',
+        'expiry_date' => 'datetime',
+        'updated_at' => 'datetime',
+        'created_at' => 'datetime',
+        'status' => 'int',
```

Comment:
it seems that is rather old:

MySQL: In version 5.0.2 (December 2003).
PostgreSQL: In version 7.4 (November 2003).
MariaDB: In version 5.1.1 (February 2009).

We could change the InvitationsMigration to use something like this

```
$table->enum('status', ['PENDING', 'ACCEPTED', 'DECLINED', 'EXPIRED', 'CANCELLED'])
                ->default('PENDING'); // Set a default value if needed
```

I have never used ENUM types as a DataBase type because of possible limitations regarding that, it is more readable though.

I also think that the underlying values of the ENUM column will start from 1 and not 0 (I am not sure if that will cause issues down the road) or we should also change the enum's values to start with 1 to be on the safe side. 

Another way would be, probably, here to use something like:

```
$table->tinyInteger('status')
                  ->unsigned()
                  ->default(InvitationStatus::PENDING)
                  ->checkIn([InvitationStatus::PENDING, InvitationStatus::ACCEPTED, InvitationStatus::DECLINED, InvitationStatus::EXPIRED, InvitationStatus::CANCELLED]);
```

It still adds numbers in the DB, but it also does a basic sanity test.

The `default` clause could be removed, based on another comment of this review..

Reply:
@defstat, that looks like a healthy baseline. Let's try with full-on database enumerations as supported by Laravel. That would be something like...

__In the enumeration class:__
```php
enum InvitationStatus: string
{
    case PENDING = 'PENDING';
    case ACCEPTED = 'ACCEPTED';
    case DECLINED = 'DECLINED';
    case EXPIRED = 'EXPIRED';
    case CANCELLED = 'CANCELLED';
}
```
__In the installation migration class:__
```php
$table->enum('status', [InvitationStatus::PENDING, InvitationStatus::ACCEPTED, InvitationStatus::DECLINED, InvitationStatus::EXPIRED, InvitationStatus::CANCELLED])
```
__In the upgrade migration class:__ (we prefer fixed values rather than references, as upgrade code needs to stay fixed over time -- but we add comments wherever we do this to aid with e.g. grepping!)
```php
$table->enum('status', ['PENDING', 'ACCEPTED, 'DECLINED, 'EXPIRED, 'CANCELLED]) // InvitationStatus::PENDING, InvitationStatus::ACCEPTED, InvitationStatus::DECLINED, InvitationStatus::EXPIRED, InvitationStatus::CANCELLED
```
This will be the first time we work with database enumeration types (and PHP enumerations too!) but I'd like to prove out this toolset so we can adopt it more thoroughly.

Reply:
@asmecher thanks - Changed.